src/code/code_0x80186A70.c

```diff
@@ -0,0 +1,307 @@
+#include <ultra64.h>
+#include <global.h>
+
+extern s32 D_801FD040[];
+extern OSIoMesg D_801FD050;
+extern OSMesgQueue D_801FD068;
+extern OSMesg D_801FD0F4;
+extern s32 D_801FD0F8;
+
+u32 osFlashGetAddr(u32 page_num) {
+    s32 temp = (D_801FD0F8 == 0) ? page_num << 6 : page_num << 7;
+
+    return temp;
+}
+
+OSPiHandle* osFlashReInit(u8 latency, u8 pulse, u8 page_size, u8 rel_duration, u32 start) {
+    D_801FD080.baseAddress = RDRAM_UNCACHED | start;
+    D_801FD080.type++;
+    D_801FD080.latency = latency;
+    D_801FD080.pulse = pulse;
+    D_801FD080.pageSize = page_size;
+    D_801FD080.relDuration = rel_duration;
+    D_801FD080.domain = 1;
+
+    return &D_801FD080;
+}
+
+void osFlashChange(u32 flash_num) {
+    D_801FD080.baseAddress = RDRAM_UNCACHED | (FRAM_STATUS_REGISTER + (flash_num << 17));
+    D_801FD080.type = 8 + flash_num;
+    return;
+}
+
+OSPiHandle* osFlashInit(void) {
+    u32 flash_type;
+    u32 flash_vendor;
+
+    osCreateMesgQueue(&D_801FD068, &D_801FD0F4, 1);
+
+    if (D_801FD080.baseAddress == (RDRAM_UNCACHED | FRAM_BASE_ADDRESS)) {
+        return &D_801FD080;
+    }
+
+    D_801FD080.type = 8;
+    D_801FD080.baseAddress = (RDRAM_UNCACHED | FRAM_BASE_ADDRESS);
+    D_801FD080.latency = 0x5;
+    D_801FD080.pulse = 0xC;
+    D_801FD080.pageSize = 0xF;
+    D_801FD080.relDuration = 0x2;
+    D_801FD080.domain = 1;
+    D_801FD080.speed = 0;
+    bzero(&D_801FD080.transferInfo, 96);
+
+    func_8008AE20(&D_801FD080);
+    osFlashReadId(&flash_type, &flash_vendor);
+
+    if (flash_vendor == 0x00C2001E || flash_vendor == 0x00C20001 || flash_vendor == 0x00C20000) {
+        D_801FD0F8 = 0; // OLD FLASH
+    } else {
+        D_801FD0F8 = 1; // NEW FLASH
+    }
+
+    return &D_801FD080;
+}
+
+void osFlashReadStatus(u8* flash_status) {
+    u32 outFlashStatus;
+
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_EXECUTE);
+    // read status using IO
+    func_800920B0(&D_801FD080, D_801FD080.baseAddress, &outFlashStatus);
+
+    // why twice ?
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_EXECUTE);
+    func_800920B0(&D_801FD080, D_801FD080.baseAddress, &outFlashStatus);
+
+    *flash_status = outFlashStatus & 0xFF;
+
+    return;
+}
+
+void osFlashReadId(u32* flash_type, u32* flash_vendor) {
+    u8 flash_status; // sp1F
+
+    // why read status ?
+    osFlashReadStatus(&flash_status);
+
+    // select silicon id read mode
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_SET_MODE_STATUS_AND_STATUS);
+
+    // read silicon id using DMA
+    D_801FD050.hdr.pri = 0;
+    D_801FD050.hdr.retQueue = &D_801FD068;
+    D_801FD050.dramAddr = D_801FD040;
+    D_801FD050.devAddr = 0;
+    D_801FD050.size = 8;
+
+    osInvalDCache(D_801FD040, 16);
```

Comment:
If this variable could be declared `extern s32 D_801FD040[4];`, then could this line be `osInvalDCache(D_801FD040, sizeof(D_801FD040));` to remove the magic number?